### ISSUE TITLE
Deleting a shared file cause vector time mismatch errors

### ISSUE THREAD

**Issue by tobous** [saros-project/saros#710]
Deleting a file during a session causes periodic errors (see stacktrace below) 
 on the host side due to vetor time mismatches.

These errors seem to stem from the [`HeartBeatDispatcher`](https://github.com/saros-project/saros/blob/master/core/src/saros/concurrent/management/HeartbeatDispatcher.java) as they no longer occur when the dispatcher is disabled.

It is not clear to me whether this is caused by the dispatcher directly (e.g. by it trying to create activities for deleted files) or whether the dispatcher is just the visible failure caused by an issue of the internal vector time handling. If it's the latter, it might be related to #223.

The issue does not seem to be critical as I couldn't observe any issues with the session after the issue occurred. Activities are still properly transferred and applied. But maybe there are issues in corner cases like when executing the recovery.

```Java
ERROR 17:23:56,178 [saros-DispatchContext] (ConcurrentDocumentServer.java:137) failed to transform jupiter activity: JupiterActivity(timestamp: [11,5], operation: Noop(0,''), source: intellij_test_1@saros-con.imp.fu-berlin.de/Saros)
saros.concurrent.jupiter.TransformationException: precondition #2 violated (Remote vector time is greater than local vector time).
	at saros.concurrent.jupiter.internal.Jupiter.checkPreconditions(Jupiter.java:265)
	at saros.concurrent.jupiter.internal.Jupiter.receiveJupiterActivity(Jupiter.java:158)
	at saros.concurrent.jupiter.internal.JupiterDocumentServer.transformJupiterActivity(JupiterDocumentServer.java:65)
	at saros.concurrent.management.JupiterServer.transform(JupiterServer.java:104)
	at saros.concurrent.management.ConcurrentDocumentServer.receive(ConcurrentDocumentServer.java:135)
	at saros.concurrent.management.ConcurrentDocumentServer.transformIncoming(ConcurrentDocumentServer.java:111)
	at saros.session.internal.ActivityHandler.directServerActivities(ActivityHandler.java:396)
	at saros.session.internal.ActivityHandler.handleIncomingActivities(ActivityHandler.java:146)
	at saros.session.internal.SarosSession.exec(SarosSession.java:730)
	at saros.session.internal.ActivitySequencer$3.run(ActivitySequencer.java:362)
	at saros.util.ThreadUtils$1.run(ThreadUtils.java:34)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

#### Steps to recreate
The error is somewhat inconsistent when just deleting a file (but it does also occur when just removing a pre-existing shared file). A consistent setup is as follows:

- Start a session on an existing module
- Create a file on the host side
- Delete the created file on host side
- Wait for at least a minute to ensure a heartbeat occurred

Works on Saros/I and Saros/E. The setup also works when creating and deleting the file on the client side.

**Comment by tobous** [comment]
I created a [release fix](https://github.com/saros-project/saros/commit/d9f65156d27ce237a3b08c1a979c6ec74093004a) just lazily disabling the heartbeat dispatcher (as it's not really necessary for Saros/I) to avoid this issue.

I am confident that this issue no longer occurs once the `HeartBeatDispatcher` is disabled.
@m273d15 Are you OK with releasing with this workaround? I have already created the draft release, uploaded the correct zip, and provided the correct branch reference to tag the correct commit.

If you don't have a problem with it, could you publish the release tomorrow morning and create the new JetBrains plugin repo entry? You will most likely be awake way before me.

**Comment by Drakulix** [comment]
This could be due to the asynchronous nature of the dispatcher, resulting in Activities for previously deleted files. But I will investigate this. Btw the `HeartbeatDispatcher` is not specific to the Saros/S, every client can theoretically suffer from the memory leak it "fixes" because the origin is the core.

Just to be sure, does this only happen in Saros/I? It should be a global issue, but I have never seen this error in Saros/E (although this might be because I have never deleted a file).

**Comment by tobous** [comment]
Yes, it also occurs with Saros/E, but its more inconsistent to get it to happen. Logs for an Eclipse session are attached below. `intellij_test_1@saros-con` is the host. The issue occurs for the first time in line 1184 in the host log and then occurs exactly every minute after that. I followed the suggested way to recreate the issue (start session, host creates file and deletes file; I also type a couple of characters on the host side in the file before deleting it), but I also had to try twice.

[Saros_intellij_test_1@saros-con.imp.fu-berlin.de_2019-09-25_20-32-47.log](https://github.com/saros-project/saros/files/3654263/Saros_intellij_test_1%40saros-con.imp.fu-berlin.de_2019-09-25_20-32-47.log)
[Saros_intellij_test_2@saros-con.imp.fu-berlin.de_2019-09-25_20-32-48.log](https://github.com/saros-project/saros/files/3654265/Saros_intellij_test_2%40saros-con.imp.fu-berlin.de_2019-09-25_20-32-48.log)

**Comment by tobous** [comment]
> Btw the `HeartbeatDispatcher` is not specific to the Saros/S, every client can theoretically suffer from the memory leak it "fixes" because the origin is the core.
> (That said, I would not discourage you to merge your "fix" for the next release as long as this is a temporary measure.)

Yeah, I know that its not specific to the server, but in most other circumstances the memory leak is very unlikely to matter as sessions are relatively short and the overhead will be minimal.

And yes, the measure is only temporary. It's on a separate branch that is only used for the release.

**Comment by srossbach** [comment]
The Heartbeat dispatcher does not work async. The worker threads syncs on the SWT thread. 

I guess the issue is that document is removed / reset on the server but it is still present on the client, so when the HeartbeatDispatcher kicks in it sends JupiterVectors for a document that should not exist on the client side but is still present.

**Comment by tobous** [comment]
From my testing, this is also resolved by #714.
@Drakulix could you test this as well to confirm it? I heard from @m273d15 that you had an approach to solve this. Is it still applicable with the other patch?
